### PR TITLE
ramp_fitting using stcal and roman_datamodels

### LINKED ISSUE #277: Update Ramp Readnoise code to RDM
_Issue [RCAL-189](https://jira.stsci.edu/browse/RCAL-189) was created on JIRA by [Paul Huwe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=phuwe):_

Update ramp steps & tests for RDM readnoise

### LINKED ISSUE #279: Update Ramp Gain Code to use RDM
_Issue [RCAL-191](https://jira.stsci.edu/browse/RCAL-191) was created on JIRA by [Paul Huwe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=phuwe):_

Update ramp steps & tests for gain RDM interface

### PR DESCRIPTION
**Pull request by nden**
Closes #279 
Closes #277 

Resolves RCAL-191
Resolves RCAL-189

**Description**

This PR implements ramp_fitting using `stcal`, `roman_datamodels` and `romancal.stpipe`.
It needs spacetelescope/stcal#43, spacetelescope/rad#66 and  spacetelescope/rad#86 .

Fixed the formatting of `CHANGES.rst`.

Checklist
- [x] Tests

- [x] Documentation

- [x ] Change log

- [ x] Milestone

- [x ] Label(s)

### CODE REVIEW

**Comment by nden** [comment]
This is a start of the ramp_fitting step working with stcal and roman_datamodels. 
Currently it fails in stcal showing a problem with the shape of the segments.
```
from romancal.ramp_fitting import ramp_fit_step
rstep=ramp_fit_step.RampFitStep()
rstep.call('jumpout.asdf', override_readnoise='rn.asdf', override_gain='gain.asdf')
reference type 'pars-rampfitstep'
2021-08-20 16:37:28,347 - stpipe.RampFitStep - INFO - RampFitStep instance created.
2021-08-20 16:37:28,389 - stpipe.RampFitStep - INFO - Step RampFitStep running with args ('jumpout.asdf',).
2021-08-20 16:37:28,390 - stpipe.RampFitStep - INFO - Step RampFitStep parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.asdf', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '', 'opt_name': '', 'maximum_cores': 'none', 'save_opt': False}
2021-08-20 16:37:28,642 - stpipe.RampFitStep - INFO - Using READNOISE reference file: /Users/dencheva/dev/romancal/rn.asdf
2021-08-20 16:37:28,660 - stpipe.RampFitStep - INFO - Using GAIN reference file: /Users/dencheva/dev/romancal/gain.asdf
2021-08-20 16:37:28,676 - stpipe.RampFitStep - INFO - Using algorithm = ols
2021-08-20 16:37:28,676 - stpipe.RampFitStep - INFO - Using weighting = optimal
2021-08-20 16:37:28,676 - stpipe.RampFitStep - WARNING - /Users/dencheva/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/stcal/ramp_fitting/ramp_fit.py:218: RuntimeWarning: invalid value encountered in sqrt
  readnoise_2d *= gain_2d / np.sqrt(2. * nframes)

2021-08-20 16:37:33,098 - stpipe.RampFitStep - INFO - data_sect shape: (7, 4096, 2112)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-df9df635fb8a> in <module>
----> 1 rstep.call('jumpout.asdf', override_readnoise='rn.asdf', override_gain='gain.asdf')

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/stpipe/step.py in call(cls, *args, **kwargs)
    588             name=name, config_file=config_file)
    589 
--> 590         return instance.run(*args)
    591 
    592     @property

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/stpipe/step.py in run(self, *args)
    405                     self.prefetch(*args)
    406                 try:
--> 407                     step_result = self.process(*args)
    408                 except TypeError as e:
    409                     if "process() takes exactly" in str(e):

~/dev/romancal/romancal/ramp_fitting/ramp_fit_step.py in process(self, input)
     58             buffsize = ramp_fit.BUFSIZE
     59 
---> 60             out_model, int_model, opt_model, gls_opt_model = ramp_fit.ramp_fit(
     61                 input_model, buffsize, self.save_opt,
     62                 readnoise_model.data, gain_model.data, self.algorithm,

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/stcal/ramp_fitting/ramp_fit.py in ramp_fit(model, buffsize, save_opt, readnoise_2d, gain_2d, algorithm, weighting, max_cores, dqflags)
    137     ramp_data = create_ramp_fit_class(model, dqflags)
    138 
--> 139     return ramp_fit_data(
    140         ramp_data, buffsize, save_opt, readnoise_2d, gain_2d,
    141         algorithm, weighting, max_cores, dqflags)

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/stcal/ramp_fitting/ramp_fit.py in ramp_fit_data(ramp_data, buffsize, save_opt, readnoise_2d, gain_2d, algorithm, weighting, max_cores, dqflags)
    219 
    220         # Compute ramp fitting using ordinary least squares.
--> 221         image_info, integ_info, opt_info = ols_fit.ols_ramp_fit_multi(
    222             ramp_data, buffsize, save_opt, readnoise_2d, gain_2d, weighting, max_cores)
    223         gls_opt_model = None

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/stcal/ramp_fitting/ols_fit.py in ols_ramp_fit_multi(ramp_data, buffsize, save_opt, readnoise_2d, gain_2d, weighting, max_cores)
    101     if number_slices == 1:
    102         # Single threaded computation
--> 103         image_info, integ_info, opt_info = ols_ramp_fit_single(
    104             ramp_data, int_times, buffsize, save_opt, readnoise_2d, gain_2d, weighting)
    105         if image_info is None:

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/stcal/ramp_fitting/ols_fit.py in ols_ramp_fit_single(ramp_data, int_times, buffsize, save_opt, readnoise_2d, gain_2d, weighting)
    634     #   saturated groups have already been flagged. The actual, fit, slopes for
    635     #   each segment are also calculated here.
--> 636     fit_slopes_ans = ramp_fit_slopes(
    637         ramp_data, gain_2d, readnoise_2d, save_opt, weighting)
    638     if fit_slopes_ans[0] == "saturated":

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/stcal/ramp_fitting/ols_fit.py in ramp_fit_slopes(ramp_data, gain_2d, readnoise_2d, save_opt, weighting)
    971             # final slopes, sigmas, etc. for the main (non-optional) products
    972             t_dq_cube, inv_var, opt_res, f_max_seg, num_seg = \
--> 973                 calc_slope(data_sect, gdq_sect, frame_time, opt_res, save_opt, rn_sect,
    974                            gain_sect, max_seg, ngroups, weighting, f_max_seg)
    975 

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/stcal/ramp_fitting/ols_fit.py in calc_slope(data_sect, gdq_sect, frame_time, opt_res, save_opt, rn_sect, gain_sect, i_max_seg, ngroups, weighting, f_max_seg)
   1655 
   1656     # Section of GROUPDQ dq section, excluding bad dq values in mask
-> 1657     gdq_sect_r = np.reshape(gdq_sect, (ngroups, npix))
   1658     mask_2d[gdq_sect_r != 0] = False  # saturated or CR-affected
   1659     mask_2d_init = mask_2d.copy()  # initial flags for entire ramp

<__array_function__ internals> in reshape(*args, **kwargs)

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/numpy/core/fromnumeric.py in reshape(a, newshape, order)
    297            [5, 6]])
    298     """
--> 299     return _wrapfunc(a, 'reshape', newshape, order=order)
    300 
    301 

~/opt/anaconda3/envs/roman-tags/lib/python3.8/site-packages/numpy/core/fromnumeric.py in _wrapfunc(obj, method, *args, **kwds)
     56 
     57     try:
---> 58         return bound(*args, **kwds)
     59     except TypeError:
     60         # A TypeError occurs if the object does have such a method in its

ValueError: cannot reshape array of size 121110528 into shape (7,8650752)

```

File: romancal/ramp_fitting/ramp_fit_step.py [hunk 1]
```diff
@@ -45,67 +39,36 @@ class RampFitStep (Step):
...
-            input_model.err = np.broadcast_to(R_input_model.err, (1,) +
-                                              R_input_model.err.shape)
-
-            input_model.refout = np.broadcast_to(R_input_model.refout, (1,) +
-                                                 R_input_model.refout.shape)
-
+        with rdd.open(input, mode='rw') as input_model:
+            input_model.data = input_model.data[np.newaxis, :]
+            input_model.data.dtype=np.float32
+            input_model.groupdq = input_model.groupdq[np.newaxis, :]
```
Comment (dmggh) [inline romancal/ramp_fitting/ramp_fit_step.py:45]: Shouldn't the err array also be 'redimensionalized' here ? ( and refout, even though it is not currently used ?)

File: romancal/ramp_fitting/ramp_fit_step.py [hunk 2]
```diff
@@ -45,67 +39,36 @@ class RampFitStep (Step):
...
             log.info('Using GAIN reference file: %s', gain_filename)
-            gain_model = datamodels.GainModel(gain_filename)
-
-            # Try to retrieve the gain factor from the gain reference file.
-            # If found, store it in the science model meta data, so that it's
-            # available later in the gain_scale step, which avoids having to
-            # load the gain ref file again in that step.
-            if gain_model.meta.exposure.gain_factor is not None:
-                input_model.meta.exposure.gain_factor = \
-                    gain_model.meta.exposure.gain_factor
```
Comment (dmggh) [inline romancal/ramp_fitting/ramp_fit_step.py:87]: I assume this is no longer needed due to DataModel refactoring ?

Comment (nden) [reply romancal/ramp_fitting/ramp_fit_step.py:87]: I was wondering - there is no `gain_factor` in the gain schema. Did we miss it or is it somewhere else? @ddavis-stsci ?

Comment (ddavis-stsci) [reply romancal/ramp_fitting/ramp_fit_step.py:87]: **Adding our conversation from slack for reference:**
dencheva:
gain models in jwst have meta.exposure.gain_factor attribute. This is missing from the roman one. Is this supposed to be in the data array for Roman? The documentation for gain reference file does not mention gain_factor.

phuwe  2:30 PM
gain_factor is in exposure in RAD

ddavis:house_with_garden:  2:35 PM
I don’t think we have gain_factors in Roman. If I recall correctly this is to account for non-standard gain factors when used in certain JWST modes. I suppose Roman can use them but so far we have no use case for a gain_factor

perry  2:36 PM
If it is needed by the existing JWST code, we can just generate a value of 1 in the converter.

ddavis:house_with_garden:  2:37 PM
The value of 1 should be fine if we need it. (edited) 

dencheva:house_with_garden:  2:48 PM
There's no need for that, it's used within the pipeline code (i.e. it's mission specific)  and in this case can be ignored. I just wanted to confirm we did not miss anything.
2:50
In jwst it's used in gain_scale. AFAIK we are not going to run gain_scale for roman but it's still listed in the docs as one of the steps
2:51
Or perhaps I misunderstand its use. @grumm may not better

grumm:slightly_smiling_face:  3:29 PM
Back in February  we thought this was needed (GH PR 120: Create gain factor keyword schema; Gain factor keyword needed. RCAL#91) . I am not aware of any more recent changes regarding whether it is needed.

**Review by ddavis-stsci — APPROVED** [review]
Looks Good.

**Pull request merged**